cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/deploy/RemoteConfigurator.java

```diff
@@ -0,0 +1,107 @@
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ *
+ */
+
+package io.cdap.cdap.internal.app.deploy;
+
+import com.google.common.util.concurrent.Futures;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.google.common.util.concurrent.SettableFuture;
+import com.google.gson.Gson;
+import io.cdap.cdap.app.deploy.ConfigResponse;
+import io.cdap.cdap.app.deploy.Configurator;
+import io.cdap.cdap.common.conf.CConfiguration;
+import io.cdap.cdap.common.conf.Constants;
+import io.cdap.cdap.common.http.DefaultHttpRequestConfig;
+import io.cdap.cdap.common.id.Id;
+import io.cdap.cdap.common.internal.remote.RemoteClient;
+import io.cdap.cdap.internal.app.deploy.pipeline.ConfiguratorConfig;
+import io.cdap.cdap.internal.app.runtime.artifact.PluginFinder;
+import io.cdap.cdap.internal.app.worker.ConfigResponseResult;
+import io.cdap.cdap.internal.app.worker.ConfiguratorTask;
+import io.cdap.cdap.internal.app.worker.RunnableTaskRequest;
+import io.cdap.common.http.HttpMethod;
+import io.cdap.common.http.HttpRequest;
+import io.cdap.common.http.HttpResponse;
+import org.apache.twill.discovery.DiscoveryServiceClient;
+import org.apache.twill.filesystem.Location;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * RemoteConfigurator sends a request to another {@link Configurator} running remotely.
+ */
+public class RemoteConfigurator implements Configurator {
+  private static final Logger LOG = LoggerFactory.getLogger(RemoteConfigurator.class);
+  private static final Gson GSON = new Gson();
+  private final ConfiguratorConfig config;
+
+  private final RemoteClient remoteClientInternal;
+
+  public RemoteConfigurator(ConfiguratorConfig config,
+                            DiscoveryServiceClient discoveryServiceClient) {
+    this.config = config;
+
+    this.remoteClientInternal = new RemoteClient(discoveryServiceClient, Constants.Service.TASK_WORKER,
+                                                 new DefaultHttpRequestConfig(false),
+                                                 String.format("%s", Constants.Gateway.INTERNAL_API_VERSION_3));
+  }
+
+  public RemoteConfigurator(CConfiguration cConf, Id.Namespace appNamespace, Id.Artifact artifactId,
+                            String appClassName, PluginFinder pluginFinder, ClassLoader artifactClassLoader,
+                            String applicationName, String applicationVersion, String configString,
+                            DiscoveryServiceClient discoveryServiceClient,
+                            Location artifactLocation) {
+
+    this.config = new ConfiguratorConfig(
+      cConf, appNamespace,
+      artifactId, appClassName,
+      applicationName,
+      applicationVersion,
+      configString, artifactLocation.toURI());
+
+    this.remoteClientInternal = new RemoteClient(discoveryServiceClient, Constants.Service.TASK_WORKER,
```

Comment:
Should call `this(config, discoverySericeClient);` instead.

Reply:
done